File: frontend/src/utils/hardwareMatcher.js
import cpuData from '../data/cpus.json'
import gpuData from '../data/gpus.json'

export function tokenize(s) {
  if (!s) return new Set()
  // Hapus symbol (r), (tm), registered, trademark, tanda kurung, koma, titik
  const clean = s.replace(/\(r\)|\(tm\)|[®™\(\),\._\.]/gi, '')
  const tokens = new Set(clean.toLowerCase().split(/[\s\-/_]+/).filter(Boolean))
  
  // Tambahkan token tanpa suffix huruf untuk model CPU/GPU (contoh: "12400f" -> "12400")
  for (const token of tokens) {
    const match = token.match(/^(\d{2,})[a-z]+$/)
    if (match) {
      tokens.add(match[1])
    }
  }
  return tokens
}

export function extractNumbers(s) {
  if (!s) return new Set()
  const clean = s
    .replace(/\(r\)|\(tm\)|[®™\(\),\._\.]/gi, '')
    .replace(/@\s*\d+(?:\.\d+)?\s*(?:ghz|mhz)/gi, '')
    .replace(/\b\d+(?:\.\d+)?\s*(?:ghz|mhz)\b/gi, '')
    .replace(/\b\d+\s*(?:gb|mb)\s*(?:vram|of vram)?\b/gi, '')
    .replace(/\b\d+(?:st|nd|rd|th)\b/gi, '')
  const matches = clean.toLowerCase().match(/\d{2,}/g)
  return new Set(matches || [])
}

function score(queryTokens, candidateName) {
  const cand = tokenize(candidateName)
  const intersection = [...queryTokens].filter(t => cand.has(t)).length
  return intersection / Math.max(queryTokens.size, cand.size)
}

export function mhzToScore(mhz) {
  const m = Number(mhz) || 0
  if (m <= 0)   return 0
  if (m < 1000) return Math.max(100, m * 0.1)
  if (m < 2000) return 1000 + (m - 1000) * 1.5
  if (m < 3000) return 2500 + (m - 2000) * 5
  if (m < 4000) return 7500 + (m - 3000) * 8
  if (m < 5000) return 15500 + (m - 4000) * 10
  return 25500 + (m - 5000) * 8
}

export function vramToScore(vram) {
  const v = Number(vram) || 0
  if (v <= 0)  return 0
  if (v < 1)   return 600
  if (v < 2)   return 1200
  if (v < 4)   return 3000
  if (v < 6)   return 6000
  if (v < 8)   return 8000
  if (v < 12)  return 12000
  if (v < 16)  return 18000
  if (v < 24)  return 25000
  return 35000
}

export function matchCpu(nameStr, fallbackMhz = 0) {
  if (!nameStr || !nameStr.trim()) {
    return { score: mhzToScore(fallbackMhz), name: 'Unknown CPU', found: false }
  }

  let resolvedMhz = fallbackMhz
  if (resolvedMhz <= 0) {
    const ghzMatch = nameStr.match(/(\d+(?:\.\d+)?)\s*ghz/i)
    if (ghzMatch) {
      resolvedMhz = Math.round(parseFloat(ghzMatch[1]) * 1000)
    } else {
      const mhzMatch = nameStr.match(/(\d+)\s*mhz/i)
      if (mhzMatch) {
        resolvedMhz = parseInt(mhzMatch[1], 10)
      }
    }
  }

  const tokens = tokenize(nameStr)
  const queryNums = extractNumbers(nameStr)
  let best = null, bestScore = 0
  for (const cpu of cpuData) {
    // Enforce model number constraint
    const candNums = extractNumbers(cpu.name)
    if (queryNums.size > 0 && candNums.size > 0) {
      const hasIntersection = [...queryNums].some(n => candNums.has(n))
      if (!hasIntersection) continue
    }

    const s = score(tokens, cpu.name)
    if (s > bestScore) { bestScore = s; best = cpu }
  }
  if (best && bestScore >= 0.3) {
    return { score: best.perf_score, name: best.name, found: true, data: best }
  }
  return { score: mhzToScore(resolvedMhz), name: nameStr, found: false }
}

export function extractVramFromName(nameStr) {
  if (!nameStr) return 0
  const sLower = nameStr.toLowerCase()
  const matchVramGb = sLower.match(/(\d+(?:\.\d+)?)\s*gb\s*(?:of\s*)?vram/)
  if (matchVramGb) {
    return parseFloat(matchVramGb[1])
  }
  const matchVramMb = sLower.match(/(\d+)\s*mb\s*(?:of\s*)?vram/)
  if (matchVramMb) {
    return Math.round((parseInt(matchVramMb[1], 10) / 1024) * 100) / 100
  }
  const matchGb = sLower.match(/(\d+(?:\.\d+)?)\s*gb/)
  if (matchGb) {
    return parseFloat(matchGb[1])
  }
  const matchMb = sLower.match(/(\d+)\s*mb/)
  if (matchMb) {
    return Math.round((parseInt(matchMb[1], 10) / 1024) * 100) / 100
  }
  if (sLower.includes("nvidia") || sLower.includes("geforce")) {
    if (["rtx 3080", "rtx 4080", "rtx 2080", "gtx 1080"].some(x => sLower.includes(x))) return 8.0
    if (["rtx 3060", "rtx 4060", "rtx 2060", "gtx 1060"].some(x => sLower.includes(x))) return 6.0
    if (["gtx 1050", "gtx 960", "gtx 760"].some(x => sLower.includes(x))) return 4.0
    if (["gtx 750", "gtx 660", "gt 1030"].some(x => sLower.includes(x))) return 2.0
    return 1.0
  }
  if (sLower.includes("radeon") || sLower.includes("amd") || sLower.includes("ati")) {
    if (sLower.includes("rx")) return 4.0
    return 1.0
  }
  if (sLower.includes("intel hd") || sLower.includes("intel iris") || sLower.includes("graphics") || sLower.includes("gma")) {
    return 0.5
  }
  return 0.0
}

export function matchGpu(nameStr, fallbackVram = 0) {
  if (!nameStr || !nameStr.trim()) {
    return { score: vramToScore(fallbackVram), name: 'Unknown GPU', found: false }
  }
  const tokens = tokenize(nameStr)
  const queryNums = extractNumbers(nameStr)
  let best = null, bestScore = 0
  for (const gpu of gpuData) {
    // Enforce model number constraint
    const candNums = extractNumbers(gpu.name)
    if (queryNums.size > 0 && candNums.size > 0) {
      const hasIntersection = [...queryNums].some(n => candNums.has(n))
      if (!hasIntersection) continue
    }

    const s = score(tokens, gpu.name)
    if (s > bestScore) { bestScore = s; best = gpu }
  }
  if (best && bestScore >= 0.3) {
    return { score: best.perf_score, name: best.name, found: true, data: best }
  }
  const resolvedVram = fallbackVram > 0 ? fallbackVram : extractVramFromName(nameStr)
  return { score: vramToScore(resolvedVram), name: nameStr, found: false }
}


export function splitConcatenatedSpecs(s) {
  if (!s) return ''
  // Sisipkan " / " sebelum kata kunci AMD, Intel, Nvidia, Radeon, GeForce, ATI, DirectX, Shader
  // yang didahului karakter, dan tidak diikuti huruf kecil (lookahead)
  let cleaned = s.replace(/([a-zA-Z0-9])(AMD|Intel|Nvidia|Radeon|GeForce|ATI|DirectX|Shader)(?![a-z])/gi, '$1 / $2')
  
  // Pisahkan model number dari VRAM yang menempel (misal: 3870512 MB -> 3870 512 MB, atau 16503 GB -> 1650 3 GB)
  cleaned = cleaned.replace(/(\d+)(1024|2048|4096|128|256|512|10|11|12|16|20|24|1|2|3|4|5|6|8)(?=\s*(MB|GB|VRAM|of))/gi, '$1 $2')

  // Pisahkan angka yang menempel di belakang kata (length >= 2 untuk hindari "i5" -> "i 5")
  cleaned = cleaned.replace(/([a-zA-Z]{2,})(\d+)/g, '$1 $2')
  return cleaned
}


export function resolveGameRequirement(cpuRaw, gpuRaw) {
  const cpuCleaned = splitConcatenatedSpecs(cpuRaw)
  const gpuCleaned = splitConcatenatedSpecs(gpuRaw)

  const splitAlts = s => (s || '').split(/\s*\/\s*|\s+or\s+/i).filter(Boolean)

  const cpuParts = splitAlts(cpuCleaned)
  const gpuParts = splitAlts(gpuCleaned)

  const cpuScores = cpuParts.map(p => matchCpu(p.trim())).filter(res => {
    // Abaikan jika tidak cocok (score 0 dan nama mengandung Unknown atau nama sama dengan input)
    if (res.score === 0 && (res.name.includes('Unknown') || !res.found)) {
      return false
    }
    return true
  }).map(res => res.score)

  const gpuScores = gpuParts.map(p => matchGpu(p.trim())).filter(res => {
    if (res.score === 0 && (res.name.includes('Unknown') || !res.found)) {
      return false
    }
    return true
  }).map(res => res.score)

  const cpuScore = cpuScores.length ? Math.min(...cpuScores) : 0
  const gpuScore = gpuScores.length ? Math.min(...gpuScores) : 0

  return { cpuScore, gpuScore }
}

export { cpuData, gpuData }

File: backend/hardware_matcher.py
"""
hardware_matcher.py
Matches arbitrary CPU/GPU name strings to the closest entry in our JSON databases.
Uses token-based scoring (no external deps).
"""
import os, json, re

_DIR = os.path.join(os.path.dirname(__file__), "data")

def _load(filename):
    path = os.path.join(_DIR, filename)
    if not os.path.exists(path):
        return []
    with open(path, encoding="utf-8") as f:
        return json.load(f)

_CPUS = None
_GPUS = None

def get_cpus():
    global _CPUS
    if _CPUS is None:
        _CPUS = _load("cpus.json")
    return _CPUS

def get_gpus():
    global _GPUS
    if _GPUS is None:
        _GPUS = _load("gpus.json")
    return _GPUS

def _tokenize(s):
    """Clean up string and return a set of normalized tokens."""
    if not s:
        return set()
    # Remove trademark symbols and common punctuation/brackets
    s_clean = re.sub(r"\(r\)|\(tm\)|[®™\(\),\._\.]", "", s, flags=re.IGNORECASE)
    # Split on spaces, dashes, slashes
    tokens = set(re.split(r"[\s\-/_]+", s_clean.lower()))
    tokens = {t for t in tokens if t}
    
    # Add extra tokens without letter suffix (e.g. "12400f" -> "12400", "3600x" -> "3600")
    extra_tokens = set()
    for token in tokens:
        match = re.match(r"^(\d{2,})[a-z]+$", token)
        if match:
            extra_tokens.add(match.group(1))
    return tokens.union(extra_tokens)

def _extract_numbers(s):
    """Find all numeric digits of length >= 2 in the string."""
    if not s:
        return set()
    s_clean = re.sub(r"\(r\)|\(tm\)|[®™\(\),\._\.]", "", s, flags=re.IGNORECASE)
    return set(re.findall(r"\d{2,}", s_clean.lower()))

def _score(query_tokens, candidate_name):
    cand_tokens = _tokenize(candidate_name)
    if not query_tokens:
        return 0
    matches = len(query_tokens & cand_tokens)
    return matches / max(len(query_tokens), len(cand_tokens))

def match_cpu(name_str, fallback_mhz=0):
    """
    Return (perf_score, matched_name).
    Falls back to a MHz-derived estimate if no match found.
    """
    if not name_str or name_str.strip() == "":
        return mhz_to_score(fallback_mhz), "Unknown CPU"

    cpus = get_cpus()
    if not cpus:
        return mhz_to_score(fallback_mhz), name_str

    tokens = _tokenize(name_str)
    query_nums = _extract_numbers(name_str)
    best, best_score = None, 0.0
    for cpu in cpus:
        # Enforce model number exact match constraint if both have model numbers
        cand_nums = _extract_numbers(cpu["name"])
        if query_nums and cand_nums:
            if not query_nums.intersection(cand_nums):
                continue

        s = _score(tokens, cpu["name"])
        if s > best_score:
            best_score = s
            best = cpu

    # Accept match if similarity >= 0.3
    if best and best_score >= 0.3:
        return best["perf_score"], best["name"]
    # fallback: use MHz heuristic
    return mhz_to_score(fallback_mhz), name_str

def extract_vram_from_name(name_str):
    if not name_str:
        return 0.0
    s_lower = name_str.lower()
    match_vram_gb = re.search(r'(\d+(?:\.\d+)?)\s*gb\s*(?:of\s*)?vram', s_lower)
    if match_vram_gb:
        return float(match_vram_gb.group(1))
    match_vram_mb = re.search(r'(\d+)\s*mb\s*(?:of\s*)?vram', s_lower)
    if match_vram_mb:
        return round(int(match_vram_mb.group(1)) / 1024.0, 2)
    match_gb = re.search(r'(\d+(?:\.\d+)?)\s*gb', s_lower)
    if match_gb:
        return float(match_gb.group(1))
    match_mb = re.search(r'(\d+)\s*mb', s_lower)
    if match_mb:
        return round(int(match_mb.group(1)) / 1024.0, 2)
    if "nvidia" in s_lower or "geforce" in s_lower:
        if any(x in s_lower for x in ["rtx 3080", "rtx 4080", "rtx 2080", "gtx 1080"]):
            return 8.0
        if any(x in s_lower for x in ["rtx 3060", "rtx 4060", "rtx 2060", "gtx 1060"]):
            return 6.0
        if any(x in s_lower for x in ["gtx 1050", "gtx 960", "gtx 760"]):
            return 4.0
        if any(x in s_lower for x in ["gtx 750", "gtx 660", "gt 1030"]):
            return 2.0
        return 1.0
    if "radeon" in s_lower or "amd" in s_lower or "ati" in s_lower:
        if "rx" in s_lower:
            return 4.0
        return 1.0
    if "intel hd" in s_lower or "intel iris" in s_lower or "graphics" in s_lower or "gma" in s_lower:
        return 0.5
    return 0.0

def match_gpu(name_str, fallback_vram=0):
    """
    Return (perf_score, matched_name).
    Falls back to a VRAM-derived estimate if no match found.
    """
    if not name_str or name_str.strip() == "":
        return vram_to_score(fallback_vram), "Unknown GPU"

    gpus = get_gpus()
    if not gpus:
        return vram_to_score(fallback_vram), name_str

    tokens = _tokenize(name_str)
    query_nums = _extract_numbers(name_str)
    best, best_score = None, 0.0
    for gpu in gpus:
        # Enforce model number exact match constraint if both have model numbers
        cand_nums = _extract_numbers(gpu["name"])
        if query_nums and cand_nums:
            if not query_nums.intersection(cand_nums):
                continue

        s = _score(tokens, gpu["name"])
        if s > best_score:
            best_score = s
            best = gpu

    if best and best_score >= 0.3:
        return best["perf_score"], best["name"]
    
    resolved_vram = fallback_vram if fallback_vram > 0 else extract_vram_from_name(name_str)
    return vram_to_score(resolved_vram), name_str


def mhz_to_score(mhz):
    """Convert raw clock speed (MHz) to approximate perf_score."""
    mhz = int(mhz or 0)
    if mhz <= 0:   return 0
    if mhz < 1000: return max(100, mhz * 0.1)
    if mhz < 2000: return 1000 + (mhz - 1000) * 1.5
    if mhz < 3000: return 2500 + (mhz - 2000) * 5
    if mhz < 4000: return 7500 + (mhz - 3000) * 8
    if mhz < 5000: return 15500 + (mhz - 4000) * 10
    return 25500 + (mhz - 5000) * 8

def vram_to_score(vram_gb):
    """Convert VRAM GB to rough GPU perf_score."""
    v = float(vram_gb or 0)
    if v <= 0:   return 0
    if v < 1:    return 600
    if v < 2:    return 1200
    if v < 4:    return 3000
    if v < 6:    return 6000
    if v < 8:    return 8000
    if v < 12:   return 12000
    if v < 16:   return 18000
    if v < 24:   return 25000
    return 35000

def split_concatenated_specs(s):
    """
    Format concatenated specifications from scraping.
    Inserts a slash separator before hardware vendor keywords and space before glued digits.
    Example: "i5-9500AMD" -> "i5-9500 / AMD"
    """
    if not s:
        return ""
    # Sisipkan " / " sebelum kata kunci AMD, Intel, Nvidia, Radeon, GeForce, ATI, DirectX, Shader
    # yang didahului karakter, dan tidak diikuti huruf kecil (lookahead)
    cleaned = re.sub(
        r'([a-zA-Z0-9])(AMD|Intel|Nvidia|Radeon|GeForce|ATI|DirectX|Shader)(?![a-z])',
        r'\1 / \2',
        s,
        flags=re.IGNORECASE
    )
    # Pisahkan model number dari VRAM yang menempel (misal: 3870512 MB -> 3870 512 MB, atau 16503 GB -> 1650 3 GB)
    cleaned = re.sub(
        r'(\d+)(1024|2048|4096|128|256|512|10|11|12|16|20|24|1|2|3|4|5|6|8)(?=\s*(MB|GB|VRAM|of))',
        r'\1 \2',
        cleaned,
        flags=re.IGNORECASE
    )
    # Pisahkan digit yang menempel di belakang kata (panjang kata >= 2 untuk menghindari "i5" -> "i 5")
    cleaned = re.sub(r'([a-zA-Z]{2,})(\d+)', r'\1 \2', cleaned)
    return cleaned

def resolve_game_requirement(cpu_raw, gpu_raw):
    """
    For a raw game requirement string (may contain alternatives like 'i5-2500K / FX-6300'),
    returns the LOWER of matched scores (minimum = can run with weaker hardware).
    """
    # Clean concatenated specifications before processing
    cpu_cleaned = split_concatenated_specs(cpu_raw or "")
    gpu_cleaned = split_concatenated_specs(gpu_raw or "")

    def pick_best_from_alts(raw, match_fn):
        parts = re.split(r"\s*/\s*|\s+or\s+", raw, flags=re.IGNORECASE)
        scores = []
        for p in parts:
            p_str = p.strip()
            if not p_str:
                continue
            score, matched_name = match_fn(p_str)
            # Abaikan noise/deskripsi alternatif yang gagal di-match (score 0 dan nama default/Unknown)
            if ("Unknown" in matched_name or matched_name == p_str) and score == 0:
                continue
            scores.append(score)
        return min(scores) if scores else 0

    cpu_score = pick_best_from_alts(cpu_cleaned, lambda s: match_cpu(s))
    gpu_score = pick_best_from_alts(gpu_cleaned, lambda s: match_gpu(s))
    return cpu_score, gpu_score
